File: problems/binary_search_trees/repair_bst/python/repair_bst.py
from problems.binary_search_trees.assets.BST import BST


def repair_bst(tree: BST) -> BST:
    first = second = prev = None

    def inorder(node):
        nonlocal first, second, prev
        if not node:
            return
        inorder(node.left)
        if prev and prev.value > node.value:
            if not first:
                first = prev
            second = node
        prev = node
        inorder(node.right)

    inorder(tree)

    if first and second:
        first.value, second.value = second.value, first.value
    return tree

File: problems/binary_search_trees/repair_bst/js/repair_bst.js
export const repairBST = (tree) => {
  let first = null,
    second = null,
    prev = null;

  const inOrder = (node) => {
    if (!node) return;
    inOrder(node.left);
    if (prev && prev.value > node.value) {
      if (!first) {
        first = prev;
      }
      second = node;
    }
    prev = node;
    inOrder(node.right);
  };
  inOrder(tree);

  if (first && second) {
    [first.value, second.value] = [second.value, first.value];
  }
  return tree;
};
